iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,307 @@
+      Type serializedVecType =
+          VectorType::get(write.getVectorType().getNumElements(),
+                          write.getVectorType().getElementType());
+      Value data = rewriter.create<vector::ShapeCastOp>(loc, serializedVecType,
+                                                        adaptor.vector());
+      data = rewriter.create<vector::BitCastOp>(
+          loc, memref.getType().cast<MemRefType>().getElementType(), data);
+      rewriter.create<StoreOp>(loc, data, memref, indices);
+    } else {
+      rewriter.create<vector::TransferWriteOp>(loc, adaptor.vector(), memref,
+                                               indices);
+    }
+    rewriter.eraseOp(write);
+    return success();
+  }
+};
+
+class ProcessAlloc final : public MemRefConversionPattern<AllocOp> {
+ public:
+  using MemRefConversionPattern<AllocOp>::MemRefConversionPattern;
+  LogicalResult matchAndRewrite(
+      AllocOp alloc, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Value, 2> dim(alloc.value());
+    unsigned elemtSize = alloc.getType().getElementTypeBitWidth();
+    unsigned vecSize = vectorizationSizeInBits / elemtSize;
+    Type vec4 = VectorType::get(4, rewriter.getF32Type());
+    SmallVector<int64_t, 2> newShape(alloc.getType().getShape().begin(),
+                                     alloc.getType().getShape().end());
+    if (newShape.back() % vecSize != 0) failure();
+    newShape.back() = newShape.back() / vecSize;
+    auto memrefType =
+        MemRefType::get(newShape, vec4, {}, alloc.getType().getMemorySpace());
+    Value newAlloc = rewriter.create<AllocOp>(alloc.getLoc(), memrefType, dim);
+    rewriter.replaceOp(alloc, newAlloc);
+    return success();
+  }
+};
+
+class VectorizeMemRefPass final
+    : public PassWrapper<VectorizeMemRefPass, OperationPass<ModuleOp>> {
+  void runOnOperation() override;
+
+ private:
+  MemRefUsageAnalysis *memrefUsageAnalysis = nullptr;
+};
+}  // namespace
+
+LogicalResult ProcessFuncArg::matchAndRewrite(
+    FuncOp funcOp, ArrayRef<Value> operands,
+    ConversionPatternRewriter &rewriter) const {
+  TypeConverter::SignatureConversion signatureConverter(
+      funcOp.getType().getNumInputs());
+  TypeConverter typeConverter;
+  for (const auto &arg : llvm::enumerate(funcOp.getArguments())) {
+    if (memrefUsageAnalysis.vectorizeMemRef(arg.value())) {
+      auto memrefType = arg.value().getType().cast<MemRefType>();
+      unsigned elemtSize = memrefType.getElementTypeBitWidth();
+      if (elemtSize % 8 != 0 || elemtSize > 32) return failure();
+      unsigned vecSize = 4 * (32 / elemtSize);
+      SmallVector<int64_t, 2> newShape(memrefType.getShape().begin(),
+                                       memrefType.getShape().end());
+      if (newShape.back() % vecSize != 0) failure();
+      newShape.back() = newShape.back() / vecSize;
+      Type vec4 = VectorType::get(4, rewriter.getF32Type());
```

Comment:
ditto, should we use vecSize instead of `4`?

Looks like we should have a method for creating this type, and it could be used in both alloc rewrite and func arg rewrite.